src/Microsoft.DotNet.XHarness.CLI/Commands/Apple/AppleOrchestrator.cs

```diff
@@ -0,0 +1,366 @@
+            TArguments arguments,
+            IEnumerable<string> passthroughArguments,
+            AppBundleInformation appBundleInfo,
+            IDevice device,
+            IDevice? companionDevice,
+            TestTargetOs target,
+            CancellationToken cancellationToken);
+
+        protected abstract Task<ExitCode> ExecuteMacCatalystApp(
+            TArguments arguments,
+            IEnumerable<string> passthroughArguments,
+            AppBundleInformation appBundleInfo,
+            CancellationToken cancellationToken);
+
+        public async Task<ExitCode> OrchestrateRun(
+            TArguments arguments,
+            IEnumerable<string> passthroughArguments,
+            TestTargetOs target,
+            CancellationToken cancellationToken)
+        {
+            _lldbFileCreated = false;
+            var isLldbEnabled = IsLldbEnabled();
+            if (isLldbEnabled && !arguments.EnableLldb)
+            {
+                // the file is present, but the user did not set it, warn him about it
+                _logger.LogWarning("Lldb will be used since '~/.mtouch-launch-with-lldb' was found in the system but it was not created by xharness.");
+            }
+            else if (arguments.EnableLldb)
+            {
+                if (!File.Exists(s_mlaunchLldbConfigFile))
+                {
+                    // create empty file
+                    File.WriteAllText(s_mlaunchLldbConfigFile, string.Empty);
+                    _lldbFileCreated = true;
+                }
+            }
+
+            var appBundleInformationParser = new AppBundleInformationParser(_processManager);
+
+            _logger.LogInformation("Getting app bundle information");
+
+            AppBundleInformation appBundleInfo;
+
+            try
+            {
+                appBundleInfo = await appBundleInformationParser.ParseFromAppBundle(
+                    arguments.AppPackagePath,
+                    target.Platform,
+                    _mainLog,
+                    cancellationToken);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError($"Failed to get bundle information: {e.Message}");
+                return ExitCode.FAILED_TO_GET_BUNDLE_INFO;
+            }
+
+            ExitCode exitCode;
+            IDevice device;
+            IDevice? companionDevice;
+
+            if (target.Platform == TestTarget.MacCatalyst)
+            {
+                try
+                {
+                    _logger.LogInformation($"Starting '{appBundleInfo.AppName}' on MacCatalyst");
+
+                    return await ExecuteMacCatalystApp(arguments, passthroughArguments, appBundleInfo, cancellationToken);
+                }
+                catch (Exception e)
+                {
+                    var message = new StringBuilder().AppendLine("Application run failed:");
+
+                    if (_errorKnowledgeBase.IsKnownTestIssue(_mainLog, out var failureMessage))
```

Comment:
would it be better to add `LogProblem` method as for `TestOrchestrator`?

Reply:
It sort of works nicely as a local function since it captures some local variables that I would have to re-declare again